core/src/main/java/org/apache/calcite/avatica/remote/looker/LookerRemoteMeta.java

```diff
@@ -0,0 +1,375 @@
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    };
+    stream.setUncaughtExceptionHandler(exceptionHandler);
+    return stream;
+  }
+
+  /**
+   * Creates a streaming iterable that parses a JSON {@link InputStream} into a series of
+   * {@link Frame}s.
+   */
+  @Override
+  public Iterable<Object> createIterable(StatementHandle h, QueryState state, Signature signature,
+      List<TypedValue> parameters, Frame firstFrame) {
+    if (LookerFrame.class.isAssignableFrom(firstFrame.getClass())) {
+      try {
+        LookerFrame lookerFrame = (LookerFrame) firstFrame;
+        String endpoint = LookerSdkFactory.queryEndpoint(lookerFrame.statementSlug);
+        // set up in/out streams
+        PipedOutputStream out = new PipedOutputStream();
+        PipedInputStream in = new PipedInputStream(out);
+        // grab the statement
+        AvaticaStatement stmt = connection.lookupStatement(h);
+        // init a new thread to stream from a Looker instance
+        Thread stream = streamingThread(endpoint, in, out);
+        stream.start();
+        // TODO bug in Avatica - the statement handle never has the signature updated
+        //  workaround by handing the signature to the iterable directly.
+        return new LookerIterable(stmt, state, firstFrame, in, signature);
+      } catch (IOException | SQLException e) {
+        throw new RuntimeException(e);
+      }
+    }
+    // if this is a normal Frame from Avatica no special treatment is needed.
+    return super.createIterable(h, state, signature, parameters, firstFrame);
+  }
+
+  /**
+   * Iterable that yields an iterator that parses an {@link InputStream} of JSON into a sequence of
+   * {@link Meta.Frame}s.
+   */
+  public class LookerIterable extends FetchIterable implements Iterable<Object> {
+
+    static final String ROWS_KEY = "rows";
+    static final String VALUE_KEY = "value";
+
+    final AvaticaStatement statement;
+    final QueryState state;
+    final Frame firstFrame;
+    final JsonParser parser;
+    final Signature signature;
+
+    LookerIterable(AvaticaStatement statement, QueryState state, Frame firstFrame, InputStream in,
+        Signature signature) {
+      super(statement, state, firstFrame);
+      assert null != in;
+      this.statement = statement;
+      this.state = state;
+      this.firstFrame = firstFrame;
+      this.signature = signature;
+      try {
+        // Create a streaming parser based on the provided InputStream
+        this.parser = new JsonFactory().createParser(in);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    @Override
+    public Iterator<Object> iterator() {
+      return new LookerIterator(statement, state, firstFrame);
+    }
+
+    /**
+     * Iterator that parses an {@link InputStream} of JSON into a sequence of {@link Meta.Frame}s.
+     */
+    public class LookerIterator extends FetchIterator implements Iterator<Object> {
```

Comment:
Have you considered running the parsing logic in the thread connected to Looker, and using a blocking queue of Frame objects (or a very similar envelope) as the buffer to invert control, converting Supplier-oriented (push) into Iterator-oriented (pull)? This structure could solve a number of problems:

1. If the queue uses an envelope, it could contain three types of objects: a bundle of rows, an exception (for transmitting those Looker connection exceptions) or an End Of Stream terminal object (eliminating the race on close).
1. The computational cost of parsing JSON and constructing rows is taken out of the client read path, reducing latency.
1. The queue's length (indirectly) controls the amount of buffering. This is slightly double-edged though, since large or wide frames may mean the queue retains reference to unexpectedly large object graphs.

Reply:
This sounds way more desirable than leaving the pipes as they are. 

So, the Looker HTTP thread would parse complete frames up to the client configured`fetchSize` (e.g. 100 rows at a time). Once a frame is complete, it is enqueued on the `BlockingQueue`. If the queue is full, the thread blocks until a frame is read off the queue.

The client would then dequeue complete frames, blocking if no frame is available. It no longer has to worry about generating these frames. 

If we are worried about point 3. we could limit the `fetchSize` to something we feel comfortable with - or even compute a max frame size based on column width + row length requested. Perhaps we can revisit that part after some testing.

Reply:
aa93bd90a48bc55d54238d13ca883aa1bd0443c0 makes the change described above using `FrameEnvelope`. The "End Of Stream terminal object" is a `Frame` object with a `done` value of `true`.